File: OmniPlay-win/src/OmniPlay.Infrastructure/Library/OmniPlayDockerClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using OmniPlay.Core.Interfaces;
using OmniPlay.Core.Models.Entities;

namespace OmniPlay.Infrastructure.Library;

public sealed class OmniPlayDockerClient : IOmniPlayDockerClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);
    private readonly HttpClient httpClient;

    public OmniPlayDockerClient(HttpClient httpClient)
    {
        this.httpClient = httpClient;
    }

    public async Task<IReadOnlyList<OmniPlayDockerLibraryItem>> GetLibraryItemsAsync(
        MediaSource source,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(source.BaseUrl, source.AuthConfig, HttpMethod.Get, "/api/library/items");
        using var response = await httpClient.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<List<OmniPlayDockerLibraryItem>>(JsonOptions, cancellationToken) ?? [];
    }

    public async Task<OmniPlayDockerLibraryDetail?> GetLibraryDetailAsync(
        MediaSource source,
        string libraryItemId,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(
            source.BaseUrl,
            source.AuthConfig,
            HttpMethod.Get,
            $"/api/library/items/{Uri.EscapeDataString(libraryItemId)}");
        using var response = await httpClient.SendAsync(request, cancellationToken);
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            return null;
        }

        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<OmniPlayDockerLibraryDetail>(JsonOptions, cancellationToken);
    }

    public Task<byte[]> GetPosterDataAsync(
        MediaSource source,
        string posterAssetId,
        CancellationToken cancellationToken = default)
    {
        return GetAssetDataAsync(source, "posters", posterAssetId, cancellationToken);
    }

    public Task<byte[]> GetThumbnailDataAsync(
        MediaSource source,
        string thumbnailAssetId,
        CancellationToken cancellationToken = default)
    {
        return GetAssetDataAsync(source, "thumbnails", thumbnailAssetId, cancellationToken);
    }

    public async Task<string?> GetPlaybackUrlAsync(
        string baseUrl,
        string? authConfig,
        string videoFileId,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(
            baseUrl,
            authConfig,
            HttpMethod.Post,
            $"/api/playback/files/{Uri.EscapeDataString(videoFileId)}/ticket");
        using var response = await httpClient.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        var ticket = await response.Content.ReadFromJsonAsync<PlaybackTicket>(JsonOptions, cancellationToken);
        return ticket?.StreamUrl is { Length: > 0 } streamUrl
            ? AbsoluteUrl(baseUrl, streamUrl)
            : null;
    }

    public async Task<OmniPlayDockerPlaybackProgress?> GetPlaybackProgressAsync(
        string baseUrl,
        string? authConfig,
        string videoFileId,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(
            baseUrl,
            authConfig,
            HttpMethod.Get,
            $"/api/playback/files/{Uri.EscapeDataString(videoFileId)}/progress");
        using var response = await httpClient.SendAsync(request, cancellationToken);
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            return null;
        }

        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<OmniPlayDockerPlaybackProgress>(JsonOptions, cancellationToken);
    }

    public async Task UpdateProgressAsync(
        string baseUrl,
        string? authConfig,
        string videoFileId,
        double positionSeconds,
        double durationSeconds,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(baseUrl, authConfig, HttpMethod.Post, "/api/playback/progress");
        request.Content = JsonContent.Create(new
        {
            videoFileId,
            positionSeconds = Math.Max(positionSeconds, 0),
            durationSeconds = Math.Max(durationSeconds, 0),
            userId = "local"
        }, options: JsonOptions);
        using var response = await httpClient.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
    }

    public async Task ImportDoubanMetadataAsync(
        MediaSource source,
        string libraryItemId,
        DoubanMetadata metadata,
        CancellationToken cancellationToken = default)
    {
        using var request = BuildRequest(
            source.BaseUrl,
            source.AuthConfig,
            HttpMethod.Post,
            $"/api/library/items/{Uri.EscapeDataString(libraryItemId)}/douban/import");
        request.Content = JsonContent.Create(new DoubanMetadataImportRequest(
            metadata.SubjectId,
            metadata.SubjectUrl,
            metadata.Title,
            metadata.OriginalTitle,
            metadata.Year,
            metadata.Rating,
            metadata.RatingCount,
            metadata.Summary,
            Genres: null,
            Countries: null,
            metadata.PosterUrl,
            metadata.FetchedAt), options: JsonOptions);

        using var response = await httpClient.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
    }

    private async Task<byte[]> GetAssetDataAsync(
        MediaSource source,
        string assetKind,
        string assetId,
        CancellationToken cancellationToken)
    {
        using var request = BuildRequest(
            source.BaseUrl,
            source.AuthConfig,
            HttpMethod.Get,
            $"/api/assets/{assetKind}/{Uri.EscapeDataString(assetId.Trim())}");
        request.Headers.Accept.Clear();
        request.Headers.Accept.ParseAdd("image/*");
        using var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
        response.EnsureSuccessStatusCode();
        var data = await response.Content.ReadAsByteArrayAsync(cancellationToken);
        return data.Length > 0
            ? data
            : throw new InvalidDataException("Docker 图片资源为空。");
    }

    private static HttpRequestMessage BuildRequest(string baseUrl, string? authConfig, HttpMethod method, string path)
    {
        var request = new HttpRequestMessage(method, AbsoluteUrl(baseUrl, path));
        request.Headers.Accept.ParseAdd("application/json");
        var config = MediaSourceAuthConfigSerializer.DeserializeOmniPlayDocker(authConfig);
        if (!string.IsNullOrWhiteSpace(config?.SessionCookie))
        {
            request.Headers.TryAddWithoutValidation("Cookie", $"omniplay_session={config.SessionCookie.Trim()}");
        }

        return request;
    }

    private static string AbsoluteUrl(string baseUrl, string value)
    {
        var normalizedBase = MediaSourceNormalizer.NormalizeBaseUrl(MediaSourceProtocol.OmniPlayDocker, baseUrl);
        if (Uri.TryCreate(value, UriKind.Absolute, out var absolute))
        {
            return absolute.ToString();
        }

        return new Uri(new Uri(normalizedBase.TrimEnd('/') + "/"), value.TrimStart('/')).ToString();
    }

    private sealed record PlaybackTicket(string StreamUrl);

    private sealed record DoubanMetadataImportRequest(
        string SubjectId,
        string SubjectUrl,
        string Title,
        string? OriginalTitle,
        string? Year,
        double? Rating,
        int? RatingCount,
        string? Summary,
        string? Genres,
        string? Countries,
        string? PosterUrl,
        DateTimeOffset FetchedAt);
}

File: OmniPlay开源-mac-x64/OmniPlay/OmniPlayDockerClient.swift
import Foundation

struct OmniPlayDockerAuthConfig: Codable {
    var username: String?
    var sessionCookie: String?

    static func encode(username: String?, sessionCookie: String?) -> String? {
        let config = OmniPlayDockerAuthConfig(
            username: username?.trimmingCharacters(in: .whitespacesAndNewlines),
            sessionCookie: sessionCookie?.trimmingCharacters(in: .whitespacesAndNewlines)
        )
        guard config.username?.isEmpty == false || config.sessionCookie?.isEmpty == false else { return nil }
        guard let data = try? JSONEncoder().encode(config) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func decode(_ value: String?) -> OmniPlayDockerAuthConfig? {
        guard let value,
              let data = value.data(using: .utf8) else {
            return nil
        }
        return try? JSONDecoder().decode(OmniPlayDockerAuthConfig.self, from: data)
    }
}

struct OmniPlayDockerLibraryItem: Decodable {
    let id: String
    let itemKind: String
    let title: String
    let releaseDate: String?
    let overview: String?
    let posterAssetId: String?
    let voteAverage: Double?
    let doubanRating: Double?
    let maxProgressSeconds: Double
    let maxDurationSeconds: Double
    let updatedAt: String
}

struct OmniPlayDockerLibraryDetail: Decodable {
    let id: String
    let itemKind: String
    let title: String
    let releaseDate: String?
    let overview: String?
    let posterAssetId: String?
    let voteAverage: Double?
    let doubanRating: Double?
    let douban: OmniPlayDockerDoubanMetadata?
    let videoFiles: [OmniPlayDockerVideoFile]
    let seasons: [OmniPlayDockerSeason]
}

struct OmniPlayDockerDoubanMetadata: Codable {
    let subjectId: String
    let subjectUrl: String
    let title: String
    let originalTitle: String?
    let year: String?
    let rating: Double?
    let ratingCount: Int?
    let summary: String?
    let genres: String?
    let countries: String?
    let posterUrl: String?
    let fetchedAt: String
}

struct OmniPlayDockerDoubanMetadataImportRequest: Encodable {
    let subjectId: String
    let subjectUrl: String
    let title: String
    let originalTitle: String?
    let year: String?
    let rating: Double?
    let ratingCount: Int?
    let summary: String?
    let genres: String?
    let countries: String?
    let posterUrl: String?
    let fetchedAt: String
}

struct OmniPlayDockerSeason: Decodable {
    let episodes: [OmniPlayDockerEpisode]
}

struct OmniPlayDockerEpisode: Decodable {
    let stillAssetId: String?
    let videoFile: OmniPlayDockerVideoFile?
    let videoFiles: [OmniPlayDockerVideoFile]

    private enum CodingKeys: String, CodingKey {
        case stillAssetId
        case videoFile
        case videoFiles
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        stillAssetId = try container.decodeIfPresent(String.self, forKey: .stillAssetId)
        videoFile = try container.decodeIfPresent(OmniPlayDockerVideoFile.self, forKey: .videoFile)
        let parsed = try container.decodeIfPresent([OmniPlayDockerVideoFile].self, forKey: .videoFiles) ?? []
        videoFiles = parsed.isEmpty ? (videoFile.map { [$0] } ?? []) : parsed
    }
}

struct OmniPlayDockerVideoFile: Decodable {
    let id: String
    let relativePath: String
    let fileName: String
    let mediaKind: String
    let fileSizeBytes: Int64?
    let durationSeconds: Double
    let positionSeconds: Double
    let isWatched: Bool
}

struct OmniPlayDockerPlaybackTicket: Decodable {
    let streamUrl: String
    let positionSeconds: Double?
    let durationSeconds: Double?
    let isWatched: Bool?
}

struct OmniPlayDockerPlaybackProgress: Decodable {
    let positionSeconds: Double
    let durationSeconds: Double
    let isWatched: Bool
}

enum OmniPlayDockerClientError: LocalizedError {
    case invalidBaseURL
    case invalidResponse
    case authenticationRequired
    case network(String)
    case requestFailed(Int, String?)

    var errorDescription: String? {
        switch self {
        case .invalidBaseURL:
            return "Docker 服务地址无效。"
        case .invalidResponse:
            return "Docker 服务响应无效。"
        case .authenticationRequired:
            return "Docker 服务需要登录，请填写用户名和密码。"
        case .network(let message):
            return "Docker 服务连接失败：\(message)"
        case .requestFailed(let status, let message):
            if let message, !message.isEmpty {
                return "Docker 服务请求失败：HTTP \(status)，\(message)"
            }
            return "Docker 服务请求失败：HTTP \(status)。"
        }
    }
}

final class OmniPlayDockerClient {
    private let baseURL: URL
    private let session: URLSession
    private(set) var sessionCookie: String?

    init(baseURLString: String, sessionCookie: String? = nil) throws {
        let normalized = MediaSourceProtocol.omniplayDocker.normalizedBaseURL(baseURLString)
        guard let url = URL(string: normalized) else { throw OmniPlayDockerClientError.invalidBaseURL }
        self.baseURL = url
        self.sessionCookie = sessionCookie?.trimmingCharacters(in: .whitespacesAndNewlines)

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 20
        configuration.timeoutIntervalForResource = 60
        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        configuration.httpCookieStorage = nil
        configuration.httpShouldSetCookies = false
        configuration.urlCache = nil
        // Docker 服务通常在局域网内，禁用系统代理/PAC，避免局域网 HTTP 请求被代理接管后报离线。
        configuration.connectionProxyDictionary = [:]
        self.session = URLSession(configuration: configuration, delegate: LocalNetworkTrustSessionDelegate.shared, delegateQueue: nil)
    }

    func login(username: String, password: String) async throws {
        var request = try makeRequest(path: "/api/auth/login", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "username": username,
            "password": password
        ])

        let (data, response) = try await send(request)
        let http = try validate(response: response, data: data, allowUnauthorized: false)
        sessionCookie = Self.sessionCookie(from: http) ?? sessionCookie
        guard let sessionCookie, !sessionCookie.isEmpty else {
            throw OmniPlayDockerClientError.invalidResponse
        }
    }

    func libraryItems() async throws -> [OmniPlayDockerLibraryItem] {
        try await get(path: "/api/library/items")
    }

    func libraryDetail(id: String) async throws -> OmniPlayDockerLibraryDetail {
        try await get(path: "/api/library/items/\(Self.escapePath(id))")
    }

    func playbackTicket(videoFileId: String) async throws -> OmniPlayDockerPlaybackTicket {
        try await post(
            path: "/api/playback/files/\(Self.escapePath(videoFileId))/ticket",
            json: nil
        )
    }

    func playbackURL(videoFileId: String) async throws -> URL {
        let ticket = try await playbackTicket(videoFileId: videoFileId)
        guard let url = absoluteURL(ticket.streamUrl) else {
            throw OmniPlayDockerClientError.invalidResponse
        }
        return url
    }

    func playbackProgress(videoFileId: String) async throws -> OmniPlayDockerPlaybackProgress {
        try await get(path: "/api/playback/files/\(Self.escapePath(videoFileId))/progress")
    }

    func updateProgress(videoFileId: String, positionSeconds: Double, durationSeconds: Double) async throws {
        try await postEmpty(path: "/api/playback/progress", json: [
            "videoFileId": videoFileId,
            "positionSeconds": max(0, positionSeconds),
            "durationSeconds": max(0, durationSeconds),
            "userId": "local"
        ])
    }

    @discardableResult
    func importDoubanMetadata(libraryItemId: String, metadata: DoubanMetadata) async throws -> OmniPlayDockerLibraryDetail {
        let request = OmniPlayDockerDoubanMetadataImportRequest(
            subjectId: metadata.subjectId,
            subjectUrl: metadata.subjectURL,
            title: metadata.title,
            originalTitle: metadata.originalTitle,
            year: metadata.year,
            rating: metadata.rating,
            ratingCount: metadata.ratingCount,
            summary: metadata.summary,
            genres: metadata.genres,
            countries: metadata.countries,
            posterUrl: metadata.posterURL,
            fetchedAt: ISO8601DateFormatter().string(from: Date(timeIntervalSince1970: metadata.fetchedAt))
        )
        return try await postEncoded(path: "/api/library/items/\(Self.escapePath(libraryItemId))/douban/import", body: request)
    }

    nonisolated func posterURL(assetId: String) -> String {
        absoluteString(path: "/api/assets/posters/\(Self.escapePath(assetId))")
    }

    func posterData(assetId: String) async throws -> Data {
        try await assetData(path: "/api/assets/posters/\(Self.escapePath(assetId))")
    }

    func thumbnailData(assetId: String) async throws -> Data {
        try await assetData(path: "/api/assets/thumbnails/\(Self.escapePath(assetId))")
    }

    private func assetData(path: String) async throws -> Data {
        for attempt in 0..<2 {
            do {
                let request = try makeRequest(path: path, method: "GET")
                let (data, response) = try await send(request)
                _ = try validate(response: response, data: data)
                guard !data.isEmpty else { throw OmniPlayDockerClientError.invalidResponse }
                return data
            } catch {
                guard attempt == 0 else { throw error }
                try await Task.sleep(nanoseconds: 250_000_000)
            }
        }
        throw OmniPlayDockerClientError.invalidResponse
    }

    private func get<T: Decodable>(path: String) async throws -> T {
        let request = try makeRequest(path: path, method: "GET")
        let (data, response) = try await send(request)
        _ = try validate(response: response, data: data)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func post<T: Decodable>(path: String, json: [String: Any]?) async throws -> T {
        var request = try makeRequest(path: path, method: "POST")
        if let json {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
        }
        let (data, response) = try await send(request)
        _ = try validate(response: response, data: data)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func postEmpty(path: String, json: [String: Any]?) async throws {
        var request = try makeRequest(path: path, method: "POST")
        if let json {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
        }
        let (data, response) = try await send(request)
        _ = try validate(response: response, data: data)
    }

    private func postEncodedEmpty<T: Encodable>(path: String, body: T) async throws {
        var request = try makeRequest(path: path, method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await send(request)
        _ = try validate(response: response, data: data)
    }

    private func postEncoded<Response: Decodable, Body: Encodable>(path: String, body: Body) async throws -> Response {
        var request = try makeRequest(path: path, method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await send(request)
        _ = try validate(response: response, data: data)
        return try JSONDecoder().decode(Response.self, from: data)
    }

    private func makeRequest(path: String, method: String) throws -> URLRequest {
        guard let url = absoluteURL(path) else { throw OmniPlayDockerClientError.invalidBaseURL }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let sessionCookie, !sessionCookie.isEmpty {
            request.setValue("omniplay_session=\(sessionCookie)", forHTTPHeaderField: "Cookie")
        }
        return request
    }

    private func send(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch let error as URLError {
            let endpoint = request.url?.absoluteString ?? "未知地址"
            throw OmniPlayDockerClientError.network(
                "\(endpoint)（\(error.code.rawValue)：\(error.localizedDescription)）。请确认 NAS 与 Mac 在同一局域网、端口 45722 已放行，并检查 Docker 服务是否正在运行。"
            )
        }
    }

    @discardableResult
    private func validate(response: URLResponse, data: Data, allowUnauthorized: Bool = false) throws -> HTTPURLResponse {
        guard let http = response as? HTTPURLResponse else {
            throw OmniPlayDockerClientError.invalidResponse
        }
        if (200...299).contains(http.statusCode) {
            return http
        }
        if http.statusCode == 401 && !allowUnauthorized {
            throw OmniPlayDockerClientError.authenticationRequired
        }
        throw OmniPlayDockerClientError.requestFailed(http.statusCode, Self.errorMessage(from: data))
    }

    nonisolated private func absoluteURL(_ value: String) -> URL? {
        if let url = URL(string: value), url.scheme != nil {
            return url
        }
        return absoluteURL(path: value)
    }

    nonisolated private func absoluteURL(path: String) -> URL? {
        guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else { return nil }
        let split = path.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        let pathPart = split.first.map(String.init) ?? path
        let queryPart = split.count > 1 ? String(split[1]) : nil
        let basePath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let relative = pathPart.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = "/" + [basePath, relative].filter { !$0.isEmpty }.joined(separator: "/")
        components.percentEncodedQuery = queryPart
        return components.url
    }

    nonisolated private func absoluteString(path: String) -> String {
        absoluteURL(path: path)?.absoluteString ?? path
    }

    private static func sessionCookie(from response: HTTPURLResponse) -> String? {
        let values = response.allHeaderFields.compactMap { key, value -> String? in
            guard String(describing: key).caseInsensitiveCompare("Set-Cookie") == .orderedSame else { return nil }
            if let value = value as? String { return value }
            if let values = value as? [String] { return values.joined(separator: ";") }
            return String(describing: value)
        }
        for value in values {
            for segment in value.components(separatedBy: ";") {
            let trimmed = segment.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed.hasPrefix("omniplay_session=") else { continue }
            let token = String(trimmed.dropFirst("omniplay_session=".count))
            return token.isEmpty ? nil : token
            }
        }
        return nil
    }

    nonisolated private static func escapePath(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }

    private static func errorMessage(from data: Data) -> String? {
        guard !data.isEmpty,
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return object["error"] as? String ?? object["message"] as? String
    }
}
